### ISSUE TITLE
Building with Debian `minimal` packaging type

### ISSUE THREAD

**Issue by dubvulture** [Eloston/ungoogled-chromium#293]
The standard way of building described in [BUILDING.md](https://github.com/Eloston/ungoogled-chromium/blob/master/BUILDING.md) has been deprecated with 73f8852.

Now `generate_build_files.py` throws and error since the "conservative" folder has been deleted (and applying the patch* does not change the directory name) so we're left with a dilemma.
Patching and renaming handled by the script? Or manually and then proceeding with the standard flavor? 

*which by the way is broken due to 29aa08f, later I'll post / pr the working one

**Comment by Eloston** [comment]
Ah, thanks for reminding me. In #248, I was going to support using a patch since it's more transparent and easier to update than replacing files.

Right now, you'll need to apply the patch to `standard` manually.

**Comment by dubvulture** [comment]
Somewhat related since it relies on the conservative build patch.

In order to build on Xenial these patches are needed to work with ninja 1.5.1
- [relax-ninja-version-requirement.patch](http://bazaar.launchpad.net/~chromium-team/chromium-browser/xenial-stable/view/head:/debian/patches/relax-ninja-version-requirement.patch)
- [no-new-ninja-flag.patch](http://bazaar.launchpad.net/~chromium-team/chromium-browser/xenial-stable/view/head:/debian/patches/no-new-ninja-flag.patch)

Should they be included in the linux_conservative patches? Is that configuration even used for any other distro besides Ubuntu 16.04?

**Comment by Eloston** [comment]
> In order to build on Xenial these patches are needed to work with ninja 1.5.1

Alright, thanks.

> Should they be included in the linux_conservative patches? Is that configuration even used for any other distro besides Ubuntu 16.04?

Yes, `linux_conservative` is supposed to be a build configuration that's generic enough to work anywhere (i.e. it has conservative build flags in contrast to `linux_dynamic`). It's supposed to work anywhere just like building regular Chromium does. On a tangential note, `linux_official` is basically `linux_conservative` but with `is_official_build=true` GN setting.

(On a more unrelated note, your questions are reminding me about the documentation refactoring...)

**Comment by Eloston** [comment]
Heads up: https://github.com/Eloston/ungoogled-chromium/commit/816ceda92319ec24734ee7cd3f567b1008387260. I think these names are an improvement.

**Comment by dubvulture** [comment]
Ok I've been trying to build the portable/minimal version and here's a quick recap:

- ninja/xenial patches are working and later I'll add them (maybe in a new folder like `/resources/patches/ubuntu/` or just `/resources/patches/ungoogled-chromium/linux/`)
- These warnings pop up during compilation:
`-Wno-user-defined-warnings` (suggests `-Wno-user-defined-literals`)
`-Wno-enum-compare-switch` (suggests `-Wno-enum-compare`)
`-Wno-unused-lambda-capture` (suggests `-Wno-unused-parameter`)
- [remove-new-flags.patch](https://github.com/Eloston/ungoogled-chromium/blob/develop/resources/patches/ungoogled-chromium/linux/remove-new-flags.patch) cannot be applied by quilt but works manually.
`dpkg-source: info: the patch has fuzz which is not allowed, or is malformed`
Anyway I'm not really sure this patch is needed. Those flags are working with *clang>=3.9* which is required for *linux_portable*

And last but not least, build fails at
```[2841/31274] LINK ./chrome_sandbox
FAILED: python "../../build/toolchain/gcc_link_wrapper.py" --output="./chrome_sandbox" -- ../../../../../../../usr/lib/llvm-3.9/bin/clang++ -pie -fPIC -Wl,-z,noexecstack -Wl,-z,now -Wl,-z,relro -Wl,-z,defs -Wl,--no-as-needed -lpthread -Wl,--as-needed -fuse-ld=gold -B../../third_party/binutils/Linux_x64/Release/bin -Wl,--threads -Wl,--thread-count=4 -Wl,--icf=all -m64 -Wl,-O1 -Wl,--gc-sections -nostdlib++ -Wl,-rpath-link=. -Wl,--disable-new-dtags -o "./chrome_sandbox" -Wl,--start-group @"./chrome_sandbox.rsp"  -Wl,--end-group   -ldl -lpthread -lrt
clang: error: unknown argument: '-nostdlib++'
[2841/31274] CXX obj/ppapi/thunk/thunk/ppb_cursor_control_thunk.o
```
(clang 3.9)

**Comment by Eloston** [comment]
> These warnings pop up during compilation:

Those warnings are fine as long as it doesn't cause problems in building or the final binary.

> Anyway I'm not really sure this patch is needed. Those flags are working with clang>=3.9 which is required for linux_portable

It's a dependency for Debian `minimal`, because it's a dependency for `standard`, because I believe it's best to use the newest `clang` possible to follow what Google does (also, I like new things). Debian `minimal` probably only needs to depend on gcc, because Debian (and I believe Ubuntu as well?) still use gcc.

To clarify ungoogled-chromium's build system a bit more, the `linux_portable` build config is independent of Debian `minimal` packaging config; in general, build configs and packaging configs should be decoupled. They're used together because they both share the same goal of compatibility (of building and running) over system dependencies. In the future, new packaging configs can be added that package `linux_portable` in different ways (e.g. different distro package formats, simple archiving, and single-file app package formats).

> And last but not least, build fails at [...]

I'm not sure what is causing that build failure; if I had to guess, maybe it's because we didn't set `use_custom_libcxx=false`. I have a feeling that there may be some additional flags from `linux_rooted` that will need to be copied over to `linux_portable`.

Do note that if `clang` just keeps causing problems, we could probably switch to `gcc`. This may also increase compatibility with older systems that have an even older clang.

**Comment by dubvulture** [comment]
Using `use_custom_libcxx=false` fixed that issue, everything compiled fine... but the browser refuses to render any page (even with --disable-gpu) and extensions crash on startup.
The only feedback I'm getting is a **segfault** somewhere much like #237 but erasing any older chromium configurations does not help. Plus, the same happens on a fresh system so I don't think it's related.

I also tried removing that flag and fixing `-nostdlib++` by applying this [patch](http://bazaar.launchpad.net/~chromium-team/chromium-browser/xenial-stable/view/head:/debian/patches/revert-clang-nostdlib%2B%2B.patch) taken from the official xenial chromium build but I got the same end result.

I guess I should have enabled debug symbols :) cya in another dozen hours when the new build finishes (using `use_custom_libcxx=false` FYI)

In the meantime I found the problem with [remove-new-flags.patch](https://github.com/Eloston/ungoogled-chromium/blob/develop/resources/patches/ungoogled-chromium/linux/remove-new-flags.patch): domains substitution breaks the patch on [line 12](https://github.com/Eloston/ungoogled-chromium/blob/0f2a8c56c32e1e678c400dd2ab4497ecca5143c4/resources/patches/ungoogled-chromium/linux/remove-new-flags.patch#L12). I guess `build/config/compiler/BUILD.gn` should be added to `domain_substitution_list`?

**Comment by Eloston** [comment]
> The only feedback I'm getting is a segfault somewhere much like #237 but erasing any older chromium configurations does not help. Plus, the same happens on a fresh system so I don't think it's related.

If the `--disable-namespace-sandbox` command-line flag fixes your problem for your segfaulting build, try applying `opensuse/fix-kernel-user-namespace-crash.patch` to your source tree and rebuild. You should be able to use the namespace sandbox after that patch is applied (as reported in ` chrome://sandbox`)

If the above doesn't fix your problem, try adding these GN flags:

```
gold_path=""
goma_dir=""
linux_use_bundled_binutils=false
use_allocator="none"
use_custom_libcxx=false
use_gconf=false
use_gnome_keyring=false
use_gold=true
use_jumbo_build=true
```

> Build with is_debug=true and symbol_level=1 failed at linking process throwing an incredible amount of undefined references.

If you still have the output, what kinds of undefined references are there?

Also, you don't need to set `is_debug=true` even if you set `symbol_level=1`. ([Source: Chromium's docs/clang.md](https://chromium.googlesource.com/chromium/src/+/62.0.3202.94/docs/clang.md#using-a-custom-clang-binary). If you want to draw ideas from that document, I wouldn't recommend setting `is_component_build=true` [since it can break things](https://chromium.googlesource.com/chromium/src/+/62.0.3202.94/docs/linux_build_instructions.md#linker-crashes))

**Comment by dubvulture** [comment]
> If the --disable-namespace-sandbox command-line flag fixes your problem

I doesn't unfortunately.

> If the above doesn't fix your problem, try adding these GN flags:

I tried and I got the same undefined references. [Here](https://gist.github.com/dubvulture/795e116a376c728ecaa555217e3eac34)'s a bunch of them.

**Comment by Eloston** [comment]
I haven't seen any behavior like that before. Maybe it's a clang issue on Ubuntu Xenial? We could try to build with GCC by setting `is_clang=false` and pull in the GCC patches from Xenial (or maybe from `resources/patches/opensuse`)

**Comment by dubvulture** [comment]
Ok here's the specific command that fails and its first undefined references:
```
FAILED: python "../../build/toolchain/gcc_solink_wrapper.py" --readelf="readelf" --nm="nm" --sofile="./libtrknotify.so" --tocfile="./libtrknotify.so.TOC" --output="./libtrknotify.so"  -- ../../../../../../../usr/lib/llvm-3.9/bin/clang++ -shared -fPIC -Wl,-z,noexecstack -Wl,-z,now -Wl,-z,relro -Wl,-z,defs -Wl,--no-as-needed -lpthread -Wl,--as-needed -fuse-ld=gold -B../../third_party/binutils/Linux_x64/Release/bin -Wl,--threads -Wl,--thread-count=4 -Wl,--icf=all -m64 -Wl,--export-dynamic -o "./libtrknotify.so" -Wl,-soname="libtrknotify.so" @"./libtrknotify.so.rsp"
../../third_party/skia/include/core/SkColor.h:33: error: undefined reference to 'SkDebugf(char const*, ...)'
../../content/public/browser/web_contents_user_data.h:55: error: undefined reference to 'content::WebContentsUserData<InfoBarService>::kLocatorKey'
../../chrome/browser/apps/app_window_registry_util.cc:25: error: undefined reference to 'g_browser_process'
../../chrome/browser/apps/app_url_redirector.cc:147: error: undefined reference to 'Profile::FromBrowserContext(content::BrowserContext*)'
../../chrome/browser/apps/app_window_registry_util.cc:27: error: undefined reference to 'ProfileManager::GetLoadedProfiles() const'
../../chrome/browser/apps/app_window_registry_util.cc:46: error: undefined reference to 'g_browser_process'
```

This happened with GCC 5.4.0 too (FYI this [patch](https://github.com/gentoo/gentoo/blob/master/www-client/chromium/files/chromium-gcc5-r3.patch) was needed for gcc)

**Comment by Eloston** [comment]
So, let me see if I'm understanding you correctly:

* With debug options enabled, both gcc and clang lead to the `libtrknotify.so` crash that you showed above.
* With debug options disabled, using GCC causes the browser to work normally (e.g. YouTube works fine)
* With debug options disabled, using clang 3.9 results in a build that partially functions (e.g. the YouTube homepage always crashes)
* You haven't tested clang 4.0 at the time of writing "EDIT2" (IMO, it probably won't be different from clang 3.9).

Are these right?

**Comment by dubvulture** [comment]
Partially. sorry for the confusion.

- Debug
  - GCC and Clang3.9 → `libtrknotify.so` crash and undefined references
- No debug
  - GCC → partially functions (e.g. the YouTube homepage always crashes)
  - Clang3.9 → fails to render any page (even internal ones like chrome://sandbox) and extensions crash upon startup

**Comment by dubvulture** [comment]
So, after the last edit I used the resulting build as my daily browser and so far I encountered no problems whatsoever.

Should I proceed with the build release?
Should I try to build without the iridium browser patchset (and what depends on them) to see if it the root of the problem? (seeing that problem with libtrknotify.so)
Just bumping the required clang version to 4.0 feels wrong to me.

**Comment by Eloston** [comment]
This is a shot in the dark, but I suspect all these behaviors are related to `libtrknotify`, and the differences are caused by the build type (Debug vs. Release) and compiler behavior.

For now, I will put this on hold. Chromium 63 is out, and Iridium updated its patches to 62.

**Comment by dubvulture** [comment]
I understand.

Anyway, I created a new [release](https://github.com/dubvulture/ungoogled-chromium-binaries/releases) with the (so far) working binaries if anyone really wants to upgrade.

**Comment by Eloston** [comment]
63 is ready for testing, so let me know how that changes things.

**Comment by dubvulture** [comment]
Clang 3.9 build results:
No problems during compilation, I just had to update some patches (pr soon)
But... every page loads correctly except for (as far as I could test)
- https://github.com/any_person/any_repo
- https://www.tensorflow.org/install/install_sources
- https://askubuntu.com/questions/000/any_question
- https://code.launchpad.net/any_package
All are throwing segfaults like the clang 3.9 chromium62 build used to do.

Debug build soon I guess.

**Comment by Eloston** [comment]
Hm, ok. I'm not sure if this affects debug symbols, but I recently added `debian/fixes/optimize.patch`.

Unrelated, but do you still need Node.js since there is `optimize_webui=false` in `common`? I don't need it in my build.

**Comment by dubvulture** [comment]
>  do you still need Node.js

Nope. (but to be honest I don't even remember if it was required before)

Anyway, **libtrknotify.so** strikes again with its failure during linking process.

**Comment by Eloston** [comment]
I tried building `linux_portable` with `linux_simple`, and everything worked on my machine for building Debian Stretch packages.

So with the current code in `develop`, here's what I did on the machine I build the Debian Stretch packages with (from the ungoogled-chromium source):

```
export UTILIKIT_CONFIG_TYPE=linux_simple
./utilikit/prepare_sources.py
./utilikit/substitute_domains.py
./utilikit/generate_build_files.py linux_simple --apply-domain-substitution
cd build/sandbox
export CLANG_BASE_PATH=/usr/lib/llvm-3.9/
./ungoogled_linux_simple/build.sh
./out/Default/chrome
```

and everything just works.

(I should note that I have user namespaces support enabled already, and I didn't build the SUID sandbox.)

**Comment by dubvulture** [comment]
Update:

- Building (**debian minimal**, linux_portable) with clang 4.0 resulted in a build that failed to render ANY page (for the record, with v62 it worked)
- Building (**linux_simple**, linux_portable) with clang 3.9 resulted in a working build.

**Comment by Eloston** [comment]
Try a54128418114c0d6d8a4e5f4530cbf0d126bde71

**Comment by dubvulture** [comment]
Tried with clang 3.9 + the new patch and it seems to be working.
I'll build again with clang 4.0 just to double check.

**Comment by Eloston** [comment]
So, it seems that the missing files caused Chromium to break. Since the different compilers resulted in varying behavior, I'm wondering if it was the EGL shared objects that are the causing the major problems. Those pages you linked in https://github.com/Eloston/ungoogled-chromium/issues/293#issuecomment-355673840 may be using may be using certain web features (such as certain CSS transforms) that cause Chromium to use 3D acceleration, thus invoking EGL.

**Comment by Eloston** [comment]
I have reverted the minimal packaging type patch, and switching the build to use LLVM 5.0 with LLD (which is available in 17.10). I will try this out and see how it goes.

**Comment by Eloston** [comment]
The standard flavor seeems to be working on 17.10, so I don't think I will try minimal this time. And since minimal worked in 63, I believe this issue's resolved already.
